libs/blocks/bulk/utils.js

```diff
@@ -0,0 +1,269 @@
+  return storedUrls ? storedUrls.join('\n') : '';
+};
+
+export const storeUrls = (urls) => {
+  setLocalStorage(BULK_STORED_URLS, urls);
+};
+
+export const storeOperation = (operation) => {
+  setLocalStorage(BULK_STORED_OPERATION, operation);
+};
+export const getUser = async () => {
+  const profile = await window.adobeIMS?.getProfile();
+  return profile
+    ? { name: profile.name, email: profile.email }
+    : { name: 'anonymous', email: 'anonymous' };
+};
+
+export const signIn = async () => {
+  // force user to sign in
+  const accessToken = await getImsToken(loadScript);
+  if (!accessToken) {
+    window.adobeIMS.signIn();
+    return false;
+  }
+  return true;
+};
+
+export const signOut = (e) => {
+  e.preventDefault();
+  window.adobeIMS?.signOut();
+};
+
+export const getAuthorizedUsers = async () => {
+  let authorizedUsers = getLocalStorage(BULK_AUTHORIZED_USERS);
+  if (authorizedUsers) return authorizedUsers;
+  const resp = await fetch(BULK_CONFIG_FILE_PATH);
+  const json = await resp.json();
+  authorizedUsers = json.users.data.map((user) => user.user);
+  setLocalStorage(BULK_AUTHORIZED_USERS, authorizedUsers);
+  return authorizedUsers;
+};
+
+const getSupportedSites = async () => {
+  let supportedSites = getLocalStorage(BULK_SUPPORTED_SITES);
+  if (supportedSites) return supportedSites;
+  const resp = await fetch(BULK_CONFIG_FILE_PATH);
+  const json = await resp.json();
+  supportedSites = json.sites.data.map((site) => site.origin);
+  setLocalStorage(BULK_SUPPORTED_SITES, supportedSites);
+  return supportedSites;
+};
+
+export const userIsAuthorized = async () => {
+  const { email } = await getUser();
+  const users = await getAuthorizedUsers();
+  return users.includes(email);
+};
+
+const siteIsSupported = async (url) => {
+  try {
+    const { origin } = new URL(url);
+    const sites = await getSupportedSites();
+    return sites.includes(origin);
+  } catch (e) {
+    return false;
+  }
+};
+
+export const getUrls = (element) => {
+  return element.current?.value.split('\n')
+    .filter((url) => url.length > 0)
+    .map((e) => e.trim());
+};
+
+export const getActionName = (action, useGerund) => {
+  let name;
+  switch (action) {
+    case null:
+    case 'preview':
+      name = (!useGerund) ? 'Preview' : 'Previewing';
+      break;
+    case 'publish':
+      name = (!useGerund) ? 'Publish' : 'Publishing';
+      break;
+    default:
+      name = (!useGerund) ? 'Preview & publish' : 'Previewing & publishing';
+  }
+  return name;
+};
+
+const executeAction = async (action, url) => {
+  const operation = (action === 'preview') ? 'preview' : 'live';
+  const siteAllowed = await siteIsSupported(url);
+  if (!siteAllowed) return UNSUPPORTED_SITE_STATUS;
+  const { hostname, pathname } = new URL(url);
+  const [branch, repo, owner] = hostname.split('.')[0].split('--');
+  const adminURL = `${ADMIN_BASE_URL}/${operation}/${owner}/${repo}/${branch}${pathname}`;
```

Comment:
On a general note, executing potentially long-running operations (in this case subsequent actions up until 1k pages) in the browser are not really considered stable. The alternative is to offload this to a microservice, e.g. I/O runtime action. Both loc and floodgate are currently refactored to move the traditional backend processing to external microservices.

Reply:
> @honstar I agree. As management and authors expect this tool to be released soon, I would release the bulk publish feature as it is now. Moving the processing to an external service should be tackled in a separate task. @rgclayton WDYT?

I agree. I'm assuming there will be some refinement as it used and we get more feedback.